### PR TITLE
Spark 3.4: Distribution and ordering enhancements

### LINKED ISSUE #7633: Improve distribution and ordering in Spark 3.4
### Feature Request / Improvement

Given that we have support for AQE writes in Spark 3.4, we should reconsider our distribution and ordering requirements.

### Query engine

Spark

### PR DESCRIPTION
**Pull request by aokolnychyi**
This fixes #7633.

### CODE REVIEW

File: core/src/main/java/org/apache/iceberg/io/BasePositionDeltaWriter.java [hunk 1]
```diff
@@ -33,14 +35,18 @@
 
   private boolean closed;
 
+  public BasePositionDeltaWriter(
```
Comment (aokolnychyi) [inline core/src/main/java/org/apache/iceberg/io/BasePositionDeltaWriter.java:38]: I will move this part into a separate PR.

File: spark/v3.4/spark-extensions/src/main/scala/org/apache/spark/sql/catalyst/analysis/RewriteMergeIntoTable.scala [hunk 2]
```diff
@@ -210,13 +206,11 @@ object RewriteMergeIntoTable extends RewriteRowLevelIcebergCommand with Predicat
     val joinHint = JoinHint(leftHint = Some(HintInfo(Some(NO_BROADCAST_HASH))), rightHint = None)
     val joinPlan = Join(NoStatsUnaryNode(targetTableProj), sourceTableProj, joinType, Some(cond), joinHint)
 
-    // add an extra matched action to output the original row if none of the actual actions matched
-    // this is needed to keep target rows that should be copied over
-    val matchedConditions = matchedActions.map(actionCondition) :+ TrueLiteral
-    val matchedOutputs = matchedActions.map(actionOutput(_, metadataAttrs)) :+ readAttrs
+    val matchedConditions = matchedActions.map(actionCondition)
```
Comment (aokolnychyi) [inline spark/v3.4/spark-extensions/src/main/scala/org/apache/spark/sql/catalyst/analysis/RewriteMergeIntoTable.scala:209]: I can, probably, submit the logic to split updates into deletes and inserts in a separate PR as well.

File: spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/source/SparkPositionDeltaWrite.java [hunk 3]
```diff
@@ -725,12 +683,24 @@ long targetDeleteFileSize() {
...
 
-    boolean fanoutWriterEnabled() {
-      return fanoutWriterEnabled;
-    }
-
     String queryId() {
       return queryId;
     }
+
+    int specIdOrdinal() {
```
Comment (aokolnychyi) [inline spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/source/SparkPositionDeltaWrite.java:690]: Will move most of changes into a separate PR.

File: core/src/main/java/org/apache/iceberg/io/BasePositionDeltaWriter.java (hunk 1 above)
Comment (aokolnychyi) [reply core/src/main/java/org/apache/iceberg/io/BasePositionDeltaWriter.java]: Done.

File: spark/v3.4/spark-extensions/src/main/scala/org/apache/spark/sql/catalyst/analysis/RewriteMergeIntoTable.scala (hunk 2 above)
Comment (aokolnychyi) [reply spark/v3.4/spark-extensions/src/main/scala/org/apache/spark/sql/catalyst/analysis/RewriteMergeIntoTable.scala]: Done.

File: spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/source/SparkPositionDeltaWrite.java (hunk 3 above)
Comment (aokolnychyi) [reply spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/source/SparkPositionDeltaWrite.java]: Done.

File: spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/SparkWriteRequirements.java [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+ * under the License.
+ */
+package org.apache.iceberg.spark;
+
+import org.apache.spark.sql.connector.distributions.Distribution;
+import org.apache.spark.sql.connector.distributions.Distributions;
+import org.apache.spark.sql.connector.expressions.SortOrder;
+
+/** A set of requirements such as distribution and ordering reported to Spark during writes. */
+public class SparkWriteRequirements {
```
Comment (aokolnychyi) [inline spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/SparkWriteRequirements.java:26]: In the future, this class would also hold the advisory partition size.

File: spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/SparkWriteUtil.java [hunk 5]
```diff
@@ -0,0 +1,247 @@
...
+import org.apache.spark.sql.connector.expressions.SortDirection;
+import org.apache.spark.sql.connector.expressions.SortOrder;
+import org.apache.spark.sql.connector.write.RowLevelOperation.Command;
+
+/**
+ * A utility that contains helper methods for working with Spark writes.
+ *
+ * <p>Note it is an evolving internal API that is subject to change even in minor releases.
+ */
+public class SparkWriteUtil {
```
Comment (aokolnychyi) [inline spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/SparkWriteUtil.java:46]: I'd recommend checking the tests first to see what changed.

File: spark/v3.4/spark/src/test/java/org/apache/iceberg/spark/TestSparkDistributionAndOrderingUtil.java [hunk 6]
```diff
@@ -91,6 +105,42 @@ public void dropTable() {
...
+  // write mode is NONE -> unspecified distribution + LOCALLY ORDER BY id, data
+  // write mode is HASH -> unspecified distribution + LOCALLY ORDER BY id, data
+  // write mode is RANGE -> ORDER BY id, data
+  //
+  // PARTITIONED BY date, days(ts) UNORDERED
+  // -------------------------------------------------------------------------
+  // write mode is NOT SET -> CLUSTER BY date, days(ts) + LOCALLY ORDER BY date, days(ts)
+  // write mode is NOT SET (fanout) -> CLUSTER BY date, days(ts) + empty ordering
+  // write mode is NONE -> unspecified distribution + LOCALLY ORDERED BY date, days(ts)
+  // write mode is NONE (fanout) -> unspecified distribution + empty ordering
```
Comment (aokolnychyi) [inline spark/v3.4/spark/src/test/java/org/apache/iceberg/spark/TestSparkDistributionAndOrderingUtil.java:131]: I disable local sort by partition columns in regular writes and in CoW operations if fanout writers enabled. If the table sort order is undefined, no need to sort records by partition columns if we are not using clustered writers.

File: spark/v3.4/spark/src/test/java/org/apache/iceberg/spark/TestSparkDistributionAndOrderingUtil.java [hunk 7]
```diff
@@ -326,16 +445,41 @@ public void testRangeWritePartitionedSortedTable() {
...
+  }
+
   // =============================================================
   // Distribution and ordering for copy-on-write DELETE operations
   // =============================================================
   //
   // UNPARTITIONED UNORDERED
   // -------------------------------------------------------------------------
-  // delete mode is NOT SET -> CLUSTER BY _file + LOCALLY ORDER BY _file, _pos
+  // delete mode is NOT SET -> CLUSTER BY _file + empty ordering
```
Comment (aokolnychyi) [inline spark/v3.4/spark/src/test/java/org/apache/iceberg/spark/TestSparkDistributionAndOrderingUtil.java:479]: Same as for appends.

File: spark/v3.4/spark/src/test/java/org/apache/iceberg/spark/TestSparkDistributionAndOrderingUtil.java [hunk 8]
```diff
@@ -346,17 +490,21 @@ public void testRangeWritePartitionedSortedTable() {
...
-  // delete mode is RANGE -> ORDER BY date, days(ts), _file, _pos
+  // delete mode is NONE (fanout) -> unspecified distribution + empty ordering
+  // delete mode is HASH -> CLUSTER BY date, days(ts) + LOCALLY ORDER BY date, days(ts)
+  // delete mode is HASH (fanout) -> CLUSTER BY date, days(ts) + empty ordering
+  // delete mode is RANGE -> ORDER BY date, days(ts)
+  // delete mode is RANGE (fanout) -> RANGE DISTRIBUTE BY date, days(ts) + empty ordering
   //
   // PARTITIONED BY date ORDERED BY id
   // -------------------------------------------------------------------------
-  // delete mode is NOT SET -> CLUSTER BY _file + LOCALLY ORDER BY date, id
```
Comment (aokolnychyi) [inline spark/v3.4/spark/src/test/java/org/apache/iceberg/spark/TestSparkDistributionAndOrderingUtil.java:356]: I am ditching clustering by `_file` in favor of clustering by partition columns for CoW operations to reduce the number of produced files. Right now, each output task may get records from various files and partitions. Hence, we produce more than needed files. This was done to avoid OOM exceptions with too large partitions. This is no longer a problem with AQE writes in Spark 3.4.

File: spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/SparkWriteUtil.java [hunk 9]
```diff
@@ -0,0 +1,247 @@
...
+
+  private static final SortOrder[] EMPTY_ORDERING = new SortOrder[0];
+  private static final SortOrder[] EXISTING_FILE_ORDERING = orderBy(FILE_PATH, ROW_POSITION);
+  private static final SortOrder[] SPEC_ID_PARTITION_ORDERING = orderBy(SPEC_ID, PARTITION);
+  private static final SortOrder[] SPEC_ID_PARTITION_FILE_ORDERING =
+      orderBy(SPEC_ID, PARTITION, FILE_PATH);
+  private static final SortOrder[] POSITION_DELETE_ORDERING =
+      orderBy(SPEC_ID, PARTITION, FILE_PATH, ROW_POSITION);
+
+  private SparkWriteUtil() {}
```
Comment (aokolnychyi) [inline spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/SparkWriteUtil.java:66]: This utility would also hold the logic to compute the advisory partition size. It is similar to the old distribution/ordering utility but uses a shorter name.

File: spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/SparkWriteConf.java [hunk 10]
```diff
@@ -262,7 +278,50 @@ private DistributionMode defaultWriteDistributionMode() {
...
+        return updateDistributionMode();
+      case MERGE:
+        return copyOnWriteMergeDistributionMode();
+      default:
+        throw new IllegalArgumentException("Unexpected command: " + command);
+    }
+  }
+
+  public SparkWriteRequirements positionDeltaRequirements(Command command) {
+    return SparkWriteUtil.positionDeltaRequirements(
```
Comment (aokolnychyi) [inline spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/SparkWriteConf.java:306]: Can't skip distribution and ordering because the spec requires position deletes to be sorted.

File: spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/source/SparkPositionDeltaWrite.java [hunk 11]
```diff
@@ -506,8 +503,7 @@ private abstract static class DeleteAndDataDeltaWriter extends BaseDeltaWriter {
         Context context) {
       this.delegate =
           new BasePositionDeltaWriter<>(
-              newInsertWriter(table, writerFactory, dataFileFactory, context),
```
Comment (aokolnychyi) [inline spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/source/SparkPositionDeltaWrite.java:509]: Using one data writer now.

File: spark/v3.4/spark/src/test/java/org/apache/iceberg/spark/TestSparkDistributionAndOrderingUtil.java [hunk 12]
```diff
@@ -1386,14 +1801,27 @@ public void testRangeCopyOnWriteMergePartitionedSortedTable() {
...
   //
-  // delete mode is NOT SET -> CLUSTER BY _spec_id, _partition + LOCALLY ORDER BY _spec_id,
-  // _partition, _file, _pos
-  // delete mode is NONE -> unspecified distribution + LOCALLY ORDER BY _spec_id, _partition, _file,
-  // _pos
-  // delete mode is HASH -> CLUSTER BY _spec_id, _partition + LOCALLY ORDER BY _spec_id, _partition,
-  // _file, _pos
+  // UNPARTITIONED (ORDERED & UNORDERED)
+  // -------------------------------------------------------------------------
+  // delete mode is NOT SET -> CLUSTER BY _spec_id, _partition, _file +
```
Comment (aokolnychyi) [inline spark/v3.4/spark/src/test/java/org/apache/iceberg/spark/TestSparkDistributionAndOrderingUtil.java:1806]: Adding `_file` to the clustering for unpartitioned tables as `spec_id`, `_partition` will be the same for all rows and deletes for the same data file may end up in multiple delete files.

File: spark/v3.4/spark/src/test/java/org/apache/iceberg/spark/TestSparkDistributionAndOrderingUtil.java [hunk 13]
```diff
@@ -1523,30 +1951,65 @@ public void testRangePositionDeltaDeletePartitionedTable() {
...
   //
-  // update mode is NOT SET -> CLUSTER BY _spec_id, _partition + LOCALLY ORDER BY _spec_id,
-  // _partition, _file, _pos
-  // update mode is NONE -> unspecified distribution + LOCALLY ORDER BY _spec_id, _partition, _file,
-  // _pos
-  // update mode is HASH -> CLUSTER BY _spec_id, _partition + LOCALLY ORDER BY _spec_id, _partition,
-  // _file, _pos
+  // UNPARTITIONED UNORDERED
+  // -------------------------------------------------------------------------
+  // update mode is NOT SET -> CLUSTER BY _spec_id, _partition, _file +
```
Comment (aokolnychyi) [inline spark/v3.4/spark/src/test/java/org/apache/iceberg/spark/TestSparkDistributionAndOrderingUtil.java:1956]: Same as DELETE.

File: spark/v3.4/spark/src/test/java/org/apache/iceberg/spark/TestSparkDistributionAndOrderingUtil.java [hunk 14]
```diff
@@ -1742,7 +2516,7 @@ public void testRangePositionDeltaUpdatePartitionedTable() {
   //                       LOCALLY ORDERED BY _spec_id, _partition, _file, _pos, date, days(ts)
   // merge mode is HASH -> CLUSTER BY _spec_id, _partition, date, days(ts) +
   //                       LOCALLY ORDER BY _spec_id, _partition, _file, _pos, date, days(ts)
-  // merge mode is RANGE -> RANGE DISTRIBUTE BY _spec_id, _partition, _file, date, days(ts) +
```
Comment (aokolnychyi) [inline spark/v3.4/spark/src/test/java/org/apache/iceberg/spark/TestSparkDistributionAndOrderingUtil.java:1745]: This may be debatable. We initially included `_file` to split the deletes if too many but AQE should be a better option we are sure we won't split deletes per partition unless needed.

File: spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/SparkWriteRequirements.java [hunk 15]
```diff
@@ -0,0 +1,46 @@
...
+
+  public static final SparkWriteRequirements EMPTY =
+      new SparkWriteRequirements(Distributions.unspecified(), new SortOrder[0]);
+
+  private final Distribution distribution;
+  private final SortOrder[] ordering;
+
+  public SparkWriteRequirements(Distribution distribution, SortOrder[] ordering) {
```
Comment (rdblue) [inline spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/SparkWriteRequirements.java:34]: Does this need to be public or can it be package-private?

File: spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/source/SparkPositionDeltaWrite.java [hunk 16]
```diff
@@ -622,8 +590,7 @@ private static class UnpartitionedDeltaWriter extends DeleteAndDataDeltaWriter {
 
     @Override
     public void update(InternalRow meta, InternalRow id, InternalRow row) throws IOException {
-      delete(meta, id);
-      delegate.update(row, dataSpec, null);
+      throw new UnsupportedOperationException("Update must be represented as delete and insert");
```
Comment (rdblue) [inline spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/source/SparkPositionDeltaWrite.java:593]: This makes me happy that we actually write updates as separate deletes and inserts.

File: spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/source/SparkPositionDeltaWrite.java [hunk 17]
```diff
@@ -567,43 +562,16 @@ public void close() throws IOException {
...
-        Context context) {
-      long targetFileSize = context.targetDataFileSize();
-
-      if (table.spec().isPartitioned()) {
-        // use a fanout writer for partitioned tables to write updates as they may be out of order
-        return new FanoutDataWriter<>(writerFactory, fileFactory, table.io(), targetFileSize);
-      } else {
-        return new ClusteredDataWriter<>(writerFactory, fileFactory, table.io(), targetFileSize);
-      }
+    private PartitioningWriter<InternalRow, DataWriteResult> newDataWriter(
```
Comment (rdblue) [inline spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/source/SparkPositionDeltaWrite.java:565]: Can you help me understand why this no longer supports the explicit fanout writer option?

I think it's because we are always adding a shuffle after the merge or update anyway so it never matters?

**Review by rdblue — APPROVED** [review]

**Comment by rdblue** [comment]
Overall this looks good to me. Thanks for taking the time to work through all the cases with AQE.

File: spark/v3.4/spark/src/test/java/org/apache/iceberg/spark/TestSparkDistributionAndOrderingUtil.java (hunk 6 above)
Comment (amogh-jahagirdar) [reply spark/v3.4/spark/src/test/java/org/apache/iceberg/spark/TestSparkDistributionAndOrderingUtil.java:131]: Curious, would it be useful to have all these distribution and ordering examples accessible in our public docs?

**Review by amogh-jahagirdar — APPROVED** [review]
Overall looks good to me, just a comment on documentation. Thanks @aokolnychyi !

File: spark/v3.4/spark/src/test/java/org/apache/iceberg/spark/TestSparkDistributionAndOrderingUtil.java (hunk 6 above)
Comment (rdblue) [reply spark/v3.4/spark/src/test/java/org/apache/iceberg/spark/TestSparkDistributionAndOrderingUtil.java:131]: @amogh-jahagirdar, yes. It would be good to have these documented.

File: spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/SparkWriteConf.java [hunk 18]
```diff
@@ -319,11 +378,12 @@ public DistributionMode positionDeltaMergeDistributionMode() {
     return DistributionMode.fromName(mergeModeName);
   }
 
-  public boolean useTableDistributionAndOrdering() {
+  private boolean ignoreTableDistributionAndOrdering() {
     return confParser
         .booleanConf()
         .option(SparkWriteOptions.USE_TABLE_DISTRIBUTION_AND_ORDERING)
         .defaultValue(SparkWriteOptions.USE_TABLE_DISTRIBUTION_AND_ORDERING_DEFAULT)
+        .negate()
```
Comment (RussellSpitzer) [inline spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/SparkWriteConf.java:386]: This makes me slightly apprehensive, since I don't like flipping the sign in the builder. I do think it's cleaner than just adding a ! to the return statement though so +1

File: spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/SparkWriteUtil.java [hunk 19]
```diff
@@ -0,0 +1,247 @@
...
+ */
+public class SparkWriteUtil {
+
+  private static final NamedReference SPEC_ID = ref(MetadataColumns.SPEC_ID);
+  private static final NamedReference PARTITION = ref(MetadataColumns.PARTITION_COLUMN_NAME);
+  private static final NamedReference FILE_PATH = ref(MetadataColumns.FILE_PATH);
+  private static final NamedReference ROW_POSITION = ref(MetadataColumns.ROW_POSITION);
+
+  private static final Expression[] FILE_CLUSTERING = clusterBy(FILE_PATH);
+  private static final Expression[] SPEC_ID_PARTITION_CLUSTERING = clusterBy(SPEC_ID, PARTITION);
```
Comment (RussellSpitzer) [inline spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/SparkWriteUtil.java:54]: Nit: Could just all this Partition_Clustering, at least in my head a partition is unique only in combination with the spec id

File: spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/SparkWriteUtil.java [hunk 20]
```diff
@@ -0,0 +1,247 @@
...
+      Table table, DistributionMode mode) {
+
+    switch (mode) {
+      case NONE:
+        return Distributions.unspecified();
+
+      case HASH:
+        if (table.spec().isUnpartitioned()) {
+          Expression[] clustering = concat(SPEC_ID_PARTITION_FILE_CLUSTERING, clustering(table));
+          return Distributions.clustered(clustering);
```
Comment (RussellSpitzer) [inline spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/SparkWriteUtil.java:156]: nit: can pull the return out of this branch statement, that way we wouldn't make two identical clustering variables in different scopes.

File: spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/SparkWriteUtil.java [hunk 21]
```diff
@@ -0,0 +1,247 @@
...
+        } else {
+          Expression[] clustering = concat(SPEC_ID_PARTITION_CLUSTERING, clustering(table));
+          return Distributions.clustered(clustering);
+        }
+
+      case RANGE:
+        if (table.spec().isUnpartitioned()) {
+          SortOrder[] ordering = concat(SPEC_ID_PARTITION_FILE_ORDERING, ordering(table));
+          return Distributions.ordered(ordering);
```
Comment (RussellSpitzer) [inline spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/SparkWriteUtil.java:165]: same comment about return statements

File: spark/v3.4/spark/src/test/java/org/apache/iceberg/spark/TestSparkDistributionAndOrderingUtil.java (hunk 6 above)
Comment (aokolnychyi) [reply spark/v3.4/spark/src/test/java/org/apache/iceberg/spark/TestSparkDistributionAndOrderingUtil.java:131]: Will add, I believe @RussellSpitzer also had some doc update PR, I am yet to check it out.

File: spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/SparkWriteConf.java (hunk 10 above)
Comment (aokolnychyi) [reply spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/SparkWriteConf.java:306]: I have a follow-up PR that adds a fanout position delete writer. It buffers deletes into a bitmap and then produces a sorted file when closed. In the future, we either remove the requirement for position deletes to be sorted or add support for Puffin delete files that would persist bitmaps.

Comment (aokolnychyi) [reply spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/SparkWriteConf.java:306]: That's beyond this release, though.

File: spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/SparkWriteRequirements.java (hunk 15 above)
Comment (aokolnychyi) [reply spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/SparkWriteRequirements.java:34]: Good idea, I'll change. The class must stay public but I can make the construct package-private.

File: spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/SparkWriteUtil.java (hunk 19 above)
Comment (aokolnychyi) [reply spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/SparkWriteUtil.java:54]: Makes sense, will change.

File: spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/source/SparkPositionDeltaWrite.java (hunk 17 above)
Comment (aokolnychyi) [reply spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/source/SparkPositionDeltaWrite.java:565]: We always add a local sort as it is required to write out position deletes. That's why we always add a sort by partition columns if the table is partitioned.

File: spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/SparkWriteUtil.java (hunk 20 above)
Comment (RussellSpitzer) [reply spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/SparkWriteUtil.java:156]: I realize looking at future code you did this to avoid exceeding line length

**Review by RussellSpitzer — APPROVED** [review]

File: spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/SparkWriteUtil.java (hunk 20 above)
Comment (aokolnychyi) [reply spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/SparkWriteUtil.java]: After I renamed the variable above like you suggested, it fits on one line now. I've inlined it.

File: spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/SparkWriteUtil.java (hunk 19 above)
Comment (aokolnychyi) [reply spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/SparkWriteUtil.java]: Changed.

File: spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/SparkWriteConf.java (hunk 18 above)
Comment (aokolnychyi) [reply spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/SparkWriteConf.java:386]: I debated it but having negation in the main logic bothered me more :)

File: spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/SparkWriteUtil.java (hunk 19 above)
[the remark this answers is not in the document]
Comment (aokolnychyi) [reply spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/SparkWriteUtil.java]: I tried it locally but it was not very obvious for ordering below. It is a bit more explicit to include all columns one by one.

**Comment by aokolnychyi** [comment]
Thanks for reviewing, @amogh-jahagirdar @rdblue @RussellSpitzer!
